File: 프로그래머스/2/42587. 프로세스/프로세스.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int solution(vector<int> priorities, int location) {    
    queue<int> q;
    for(const auto& a : priorities) q.emplace(a);
    sort(priorities.begin(), priorities.end(), greater<int>());
    int answer = 0;
    while(true){
        if(q.front() == priorities[0]){
            answer++;
            q.pop();
            priorities.erase(priorities.begin());
            if(location == 0) return answer;
            else location--;
        }
        else{
            q.emplace(q.front());
            q.pop();
            location--;
            if(location < 0) location = q.size() - 1;
        }
    }
    return answer;
}

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    class Node implements Comparable<Node>{
        int pri;
        int loc;
        Node (int pri, int loc) {
            this.pri = pri;
            this.loc = loc;
        }
        @Override
        public int compareTo(Node n) {
            return this.pri - n.pri;
        }
    }
        
        
    public int solution(int[] priorities, int location) {
        int answer = 0;
        Deque<Node> deque = new ArrayDeque<>();
        Integer[] priorities2 = new Integer[priorities.length];
        for (int i = 0; i < priorities.length; i++) {
            deque.addLast(new Node(priorities[i], i));
            priorities2[i] = priorities[i];
        }        
        Arrays.sort(priorities2, Collections.reverseOrder());
        
        int idx = 0;
        while(idx < priorities2.length) {
            Node now = deque.removeFirst();
            if (now.pri == priorities2[idx]) {
                idx++;
                if (now.loc == location) {
                    answer = idx;
                    break;
                }
            } else {
                deque.addLast(now);
            }    
        }

        return answer;
    }
}